File: patterns/01-design-patterns/02-structural/07-flyweight/go/flyweight.go
package flyweight

type PatternError struct{ code string }

func (e *PatternError) Error() string { return e.code }
func (e *PatternError) Code() string  { return e.code }

type ModelMetadata struct {
	Model, Provider string
	ContextWindow   int
}
type Factory struct{ values map[string]*ModelMetadata }

func NewFactory() *Factory { return &Factory{values: map[string]*ModelMetadata{}} }
func (f *Factory) Register(value map[string]any) error {
	candidate := &ModelMetadata{value["model"].(string), value["provider"].(string), int(value["context_window"].(float64))}
	if current, ok := f.values[candidate.Model]; ok && (current.Provider != candidate.Provider || current.ContextWindow != candidate.ContextWindow) {
		return &PatternError{code: "intrinsic_conflict"}
	}
	f.values[candidate.Model] = candidate
	return nil
}
func (f *Factory) Get(model string) (*ModelMetadata, error) {
	value, ok := f.values[model]
	if !ok {
		return nil, &PatternError{code: "unknown_model"}
	}
	return value, nil
}
func Evaluate(input map[string]any) (any, error) {
	factory := NewFactory()
	for _, raw := range input["definitions"].([]any) {
		if err := factory.Register(raw.(map[string]any)); err != nil {
			return nil, err
		}
	}
	rawRoutes, _ := input["routes"].([]any)
	routes := make([]map[string]any, 0, len(rawRoutes))
	values := make([]*ModelMetadata, 0, len(rawRoutes))
	for _, raw := range rawRoutes {
		route := raw.(map[string]any)
		metadata, err := factory.Get(route["model"].(string))
		if err != nil {
			return nil, err
		}
		values = append(values, metadata)
		routes = append(routes, map[string]any{"model": metadata.Model, "provider": metadata.Provider, "context_window": metadata.ContextWindow, "tenant": route["tenant"]})
	}
	reused := len(values) > 1
	if reused {
		for _, value := range values[1:] {
			if value != values[0] {
				reused = false
				break
			}
		}
	}
	return map[string]any{"routes": routes, "flyweight_count": len(factory.values), "reused": reused}, nil
}

File: patterns/01-design-patterns/02-structural/07-flyweight/python/flyweight.py
from dataclasses import dataclass
class PatternError(Exception):
    def __init__(self,code):super().__init__(code);self.code=code
@dataclass(frozen=True)
class ModelMetadata:model:str;provider:str;context_window:int
class FlyweightFactory:
    def __init__(self):self.values={}
    def register(self,value):
        candidate=ModelMetadata(value["model"],value["provider"],value["context_window"]);current=self.values.get(candidate.model)
        if current is not None and current!=candidate:raise PatternError("intrinsic_conflict")
        self.values[candidate.model]=candidate
    def get(self,model):
        if model not in self.values:raise PatternError("unknown_model")
        return self.values[model]
def evaluate(input_data):
    factory=FlyweightFactory()
    for value in input_data.get("definitions",[]):factory.register(value)
    flyweights=[];routes=[]
    for route in input_data.get("routes",[]):
        metadata=factory.get(route["model"]);flyweights.append(metadata);routes.append({"model":metadata.model,"provider":metadata.provider,"context_window":metadata.context_window,"tenant":route["tenant"]})
    reused=len(flyweights)>1 and all(value is flyweights[0] for value in flyweights)
    return{"routes":routes,"flyweight_count":len(factory.values),"reused":reused}
